File: 1º Ano/2º Semestre/Programação Orientada a Objetos/Labs/Lab06/CardGames/src/CardGames/Deck.java

package CardGames;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.Stack;

/**
 * Esta classe representa um baralho de cartas
 * 
 * @author POO 2019/2020
 * @version maio/2020
 */

public class Deck {

    //private List<Card> cards;
    Stack<Card> cards;
    private Random random = new Random();

    public Deck() {
        cards = new Stack<>();
    }

    public Deck(List<Card> cardsList) {
        //this.cards = new ArrayList<>(cards);
        this.cards = new Stack<Card>();
        this.cards.addAll(cardsList);
    }

    public void addCard(Card card) {
        if (card != null) {
            cards.add(card);
        }
    }

    public boolean removeCard(Card card) {
        return cards.remove(card);
    }

    public void clear() {
        cards.clear();
    }

    @Override
    public String toString() {
        String cardsList = "";

        for (Card card : cards) {
            cardsList += card + "\n";
        }

        return cardsList;
    }

    public Card getRandomCard() {
        if (cards.isEmpty()) {
            return null;
        }
        int randomIndex = random.nextInt(cards.size() - 1);
        return cards.remove(randomIndex);
    }
    
    public void sortByValue(){
        Collections.sort(cards, Collections.reverseOrder());
    }
    
    public Iterator<Card> iterator() {
        return cards.iterator();     
    }
    
    public void shuffle() {
        Collections.shuffle(cards);
    }
    
    public List<Card> getCards() {
        ArrayList<Card> newCards = new ArrayList<Card>(this.cards);        
        return newCards;
    }
    
    public Card topCard() {
        return this.cards.firstElement();
    }
    
    public Card drawCard() {
        return this.cards.pop();
    }
    
    public void putCard(Card card) {
        this.cards.push(card);
    }
    
    public void sort(){
        Collections.sort(this.cards, new CardSuitValueComparator()); 
    }
}

File: 3º Ano/Anual/Programação Visual (PV)/Labs/Lab 1/Lab1/Deck.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Lab1
{
    class Deck
    {
        private List<Card> cards;
        private Random random = new Random();

        public Deck()
        {
            cards = new List<Card>();
        }
        public Deck(List<Card> cards)
        {
            this.cards = new List<Card>();
            if (cards != null)
            {
                foreach (Card card in cards)
                {
                    this.cards.Add(card);
                }
            }
        }

        public void AddCard(Card card)
        {
            if (card != null)
            {
                cards.Add(card);
            }
        }
        public bool RemoveCard(Card card)
        {
            return cards.Remove(card);
        }

        public void Clear()
        {
            cards.Clear();
        }

        public bool hasNext(int index)
        {
            if (this.cards[index + 1] != null)
            {
                return true;
            }
            return false;
        }
        public Card Next(int index)
        {
            if (hasNext(index))
            {
                return this.cards[index + 1];
            }
            return null;
        }

        public void Remove()
        {

        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder(" ", 100);
            foreach (Card card in cards)
            {
                sb.Append(card.ToString());
            }

            return sb.ToString();
        }
        public Card GetRandomCard()
        {
            if (cards == null)
            {
                return null;
            }
            int randomIndex = random.Next(cards.Count - 1);
            Card card = cards[randomIndex];
            cards.RemoveAt(randomIndex);
            return card;
        }

        public void SortByValue()
        {
            cards.Sort();
        }

        public void Shuffle()
        {
            Random rnd = new Random();
            for (int i=0; i < this.cards.Count; i++)
            {
                int r = rnd.Next(0, i);
                Card cardValue = cards[r];
                cards[r] = cards[i];
                cards[i] = cards[r];
            }
        }

        public List<Card> Cards
        {
            get
            {
                return new List<Card>(this.cards);
            }
        }
        public Card TopCard()
        {
            return cards[cards.Count - 1];
        }

        public Card DrawCard()
        {
            Card card = cards[cards.Count - 1];
            cards.Remove(card);
            return card;
        }
        public Card PutCard(Card card)
        {
            AddCard(card);
            return card;
        }
    }
}
